src/turbomind/models/llama/LlamaWeight.cc

```diff
@@ -88,9 +101,38 @@ LlamaWeight<T>::~LlamaWeight()
 template<typename T>
 void LlamaWeight<T>::mallocWeights()
 {
-    deviceMalloc((T**)&pre_decoder_embedding_table, vocab_size_padded_ * hidden_units_);
+    FT_CHECK(vocab_size_padded_ % tensor_para_size_ == 0);
+    deviceMalloc((T**)&pre_decoder_embedding_table, embedding_table_size_);
     deviceMalloc((T**)&output_norm_weight, hidden_units_);
-    deviceMalloc((T**)&post_decoder_embedding_kernel, hidden_units_ * vocab_size_padded_);
+    deviceMalloc((T**)&post_decoder_embedding_kernel, hidden_units_ * vocab_size_padded_ / tensor_para_size_);
+}
+
+template<typename T>
+void loadLinearWeights(T*             weights,
+                       std::string    prefix,
+                       int            rank,
+                       size_t         tensor_para_size,
+                       size_t         dim0,
+                       size_t         dim1,
+                       FtCudaDataType type,
+                       size_t         split_dim)
+{
+    FT_CHECK(split_dim == 0 || split_dim == 1);
+    // the weight could be split along split_dim
+    std::vector<std::reference_wrapper<size_t>> dims = {dim0, dim1};
+    if (dims[split_dim] % tensor_para_size == 0) {
```

Comment:
I think it's better to read the weight by rank. If the`prefix + "." + rank + ".weight" doesn't exist, fallback to `prefix + ".weight"`